sdk/clientcore/core/src/main/java/io/clientcore/core/util/Union.java

```diff
@@ -0,0 +1,352 @@
+ *
+ */
+public final class Union {
+    private static final ClientLogger LOGGER = new ClientLogger(Union.class);
+
+    private final List<Type> types;
+    private Object value;
+    private Type currentType;
+
+    private Union(Type... types) {
+        if (types == null || types.length == 0) {
+            throw LOGGER.logThrowableAsError(new IllegalArgumentException("types cannot be null or empty"));
+        }
+
+        ArrayList<Type> typeCopy = new ArrayList<>(types.length);
+        for (int i = 0; i < types.length; i++) {
+            final Type currentType = types[i];
+            if (currentType == null) {
+                throw LOGGER.logThrowableAsError(
+                    new IllegalArgumentException("types cannot contain null values: null value in index " + i));
+            } else if (!(currentType instanceof Class<?> || currentType instanceof ParameterizedType)) {
+                throw LOGGER.logThrowableAsError(new IllegalArgumentException(
+                    String.format("types must be of type Class or ParameterizedType: type name is %s in index %d.",
+                        currentType.getTypeName(), i)));
+            }
+
+            typeCopy.add(types[i]);
+        }
+        this.types = Collections.unmodifiableList(typeCopy);
+    }
+
+    /**
+     * Creates a new instance of {@link Union} with the provided types.
+     * <p>
+     * Currently, the types can be of type {@link Class} or {@link ParameterizedType}. If the type is a {@link Class},
+     * it represents a simple type. If the type is a {@link ParameterizedType}, it represents a generic type.
+     * For example, {@code List<String>} would be represented as {@code new GenericParameterizedType(List.class, String.class)}.
+     * </p>
+     *
+     * It throws {@link IllegalArgumentException} if:
+     * <ul>
+     *   <li>value is not of one of the types in the union,</li>
+     *   <li>value is null,</li>
+     *   <li>types array is null or empty,</li>
+     *   <li>types array contains a null value,</li>
+     *   <li>types array contains a type that is not of type {@link Class} or {@link ParameterizedType}.</li>
+     * </ul>
+     *
+     * @param types The types of the union.
+     * @return A new instance of {@link Union}.
+     */
+    public static Union ofTypes(Type... types) {
+        return new Union(types);
+    }
+
+    /**
+     * Sets the value of the union. A new updated immutable union is returned.
+     *
+     * @param value The value of the union.
+     * @return A new updated immutable union.
+     * @throws IllegalArgumentException If the value is not of one of the types in the union.
+     */
+    @SuppressWarnings("unchecked")
+    public Union setValue(Object value) {
+        for (Type type : types) {
+            if (isInstanceOfType(value, type) || isPrimitiveTypeMatch(value, type)) {
+                Union newUnion = new Union(types.toArray(new Type[0]));
```

Comment:
We should add an internal constructor that accepts `List<Type>` rather than `Type...`. This internal constructor should also perform no validation.

If `Union.setValue` is called often this could add a tremendous amount of overhead for repeating work we've already done.